xds/internal/xdsclient/transport/loadreport.go

```diff
@@ -0,0 +1,256 @@
+/*
+ *
+ * Copyright 2022 gRPC authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package transport
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"io"
+	"time"
+
+	"github.com/golang/protobuf/ptypes"
+	"google.golang.org/grpc/internal/pretty"
+	"google.golang.org/grpc/xds/internal"
+	"google.golang.org/grpc/xds/internal/xdsclient/load"
+	"google.golang.org/protobuf/proto"
+
+	v3corepb "github.com/envoyproxy/go-control-plane/envoy/config/core/v3"
+	v3endpointpb "github.com/envoyproxy/go-control-plane/envoy/config/endpoint/v3"
+	v3lrsgrpc "github.com/envoyproxy/go-control-plane/envoy/service/load_stats/v3"
+	v3lrspb "github.com/envoyproxy/go-control-plane/envoy/service/load_stats/v3"
+)
+
+const clientFeatureLRSSendAllClusters = "envoy.lrs.supports_send_all_clusters"
+
+type lrsStream v3lrsgrpc.LoadReportingService_StreamLoadStatsClient
+
+// ReportLoad starts reporting loads to the management server the transport is
+// configured to talk to.
+//
+// It returns a Store for the user to report loads and a function to cancel the
+// load reporting.
+func (t *Transport) ReportLoad() (*load.Store, func()) {
+	t.lrsMu.Lock()
+	defer t.lrsMu.Unlock()
+
+	t.lrsStartStream()
+	return t.lrsStore, func() {
+		t.lrsMu.Lock()
+		t.lrsStopStream()
+		t.lrsMu.Unlock()
+	}
+}
+
+// lrsStartStream starts an LRS stream to the server, if none exists.
+//
+// Caller must hold t.lrsMu.
+func (t *Transport) lrsStartStream() {
+	t.lrsRefCount++
+	if t.lrsRefCount != 1 {
+		// Return early if the stream has already been started.
+		return
+	}
+
+	ctx, cancel := context.WithCancel(context.Background())
+	t.lrsCancelStream = cancel
+	go t.reportLoad(ctx)
+}
+
+// lrsStopStream closes the LRS stream, if this is the last user of the stream.
+//
+// Caller must hold t.lrsMu.
+func (t *Transport) lrsStopStream() {
+	t.lrsRefCount--
+	if t.lrsRefCount != 0 {
+		// Return early if the stream has other references.
+		return
+	}
+
+	t.lrsCancelStream()
+	t.logger.Infof("Stopping LRS stream")
```

Comment:
I think in principle this should block until the stream has exited.  This may be worth doing now.